code/datums/components/swabbing.dm

```diff
@@ -0,0 +1,114 @@
+/*!
+
+This component is used in vat growing to swab for microbiological samples which can then be mixed with reagents in a petridish to create a culture plate.
+
+*/
+/datum/component/swabbing
+	///The current datums on the swab
+	var/list/swabbed_items
+	///Can we swab objs?
+	var/CanSwabObj
+	///Can we swab turfs?
+	var/CanSwabTurf
+	///Can we swab mobs?
+	var/CanSwabMob
+	///Callback for update_icon()
+	var/datum/callback/UpdateIcons
+	///Callback for update_overlays()
+	var/datum/callback/UpdateOverlays
+
+/datum/component/swabbing/Initialize(CanSwabObj = TRUE, CanSwabTurf = TRUE, CanSwabMob = FALSE, datum/callback/UpdateIcons, datum/callback/UpdateOverlays, swab_time = 10, max_items = 3)
+	if(!isitem(parent))
+		return COMPONENT_INCOMPATIBLE
+
+	RegisterSignal(parent, COMSIG_ITEM_PRE_ATTACK, .proc/TryToSwab)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/handle_overlays)
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_ICON, .proc/handle_icon)
+
+	src.CanSwabObj = CanSwabObj
+	src.CanSwabTurf = CanSwabTurf
+	src.CanSwabMob = CanSwabMob
+	src.UpdateIcons = UpdateIcons
+	src.UpdateOverlays = UpdateOverlays
+
+///Changes examine based on your sample
+/datum/component/swabbing/proc/examine(datum/source, mob/user, list/examine_list)
+	if(LAZYLEN(swabbed_items))
+		examine_list += "<span class='nicegreen'>There is a microbiological sample on [parent]!</span>"
+		examine_list += "<span class='notice'>You can see the following micro-organisms:</span>\n"
+		for(var/i in swabbed_items)
+			var/datum/biological_sample/samp = i
+			examine_list += samp.GetAllDetails(user.research_scanner) //Get just the names nicely parsed.
+
+///Ran when you attack an object, tries to get a swab of the object. if a swabbable surface is found it will run behavior and hopefully
+/datum/component/swabbing/proc/TryToSwab(datum/source, atom/target, mob/user, params)
+	set waitfor = FALSE //This prevents do_after() from making this proc not return it's value.
+
+	if(istype(target, /obj/structure/table))//help how do i do this less shitty
+		return NONE //idk bro pls send help
+
+	if(istype(target, /obj/item/petri_dish))
+		if(!LAZYLEN(swabbed_items))
+			return NONE
+		var/obj/item/petri_dish/dish = target
+		if(dish.sample)
+			return
+
+		var/datum/biological_sample/deposited_sample
+
+		for(var/datum/biological_sample/S in swabbed_items) //Typed in case there is a non sample on the swabbing tool because someone was fucking with swabbable element
```

Comment:
Add a type check inside, and make it scream an error with a `stack_trace()` and `continue` if there is.
Or just let it runtime if someone does bad code.
We want to avoid silent errors.

Reply:
ok